pkg/render/node.go

```diff
@@ -520,62 +514,112 @@ func (c *nodeComponent) nodeCNIConfigMap() *corev1.ConfigMap {
-		portmap = `,
-    {"type": "portmap", "snat": true, "capabilities": {"portMappings": true}}`
-	}
-
 	ipam := c.getCalicoIPAM()
 	if c.cfg.Installation.CNI.IPAM.Type == operatorv1.IPAMPluginHostLocal {
 		ipam = buildHostLocalIPAM(c.cfg.Installation.CalicoNetwork)
 	}
 
-	var k8sAPIRoot string
 	apiRoot := c.cfg.K8sServiceEp.CNIAPIRoot()
+
+	// Determine logging configuration
+	logSeverity := string(*c.cfg.Installation.Logging.CNILogging.LogSeverity)
+
+	// store logFileMaxSize as megabyte in config map
+	logFileMaxSize := c.cfg.Installation.Logging.CNILogging.LogFileMaxSize.Value() / (1024 * 1024)
+
+	logFileMaxAge := *c.cfg.Installation.Logging.CNILogging.LogFileMaxAgeDays
+
+	logFileMaxCount := *c.cfg.Installation.Logging.CNILogging.LogFileMaxCount
+
+	// calico plugin
+	var calicoPlugin = map[string]interface{}{
+		"type":                   "calico",
+		"datastore_type":         "kubernetes",
+		"mtu":                    mtu,
+		"nodename_file_optional": nodenameFileOptional,
+		"log_level":              logSeverity,
+		"log_file_path":          "/var/log/calico/cni/cni.log",
+		"log_file_max_size":      logFileMaxSize,
+		"log_file_max_age":       logFileMaxAge,
+		"log_file_max_count":     logFileMaxCount,
+		"ipam":                   ipam,
+		"container_settings": map[string]interface{}{
+			"allow_ip_forwarding": ipForward,
+		},
+		"policy": map[string]interface{}{
+			"type": "k8s",
+		},
+	}
+
+	// optional properties
+	var kubernetes = map[string]interface{}{
+		"kubeconfig": "__KUBECONFIG_FILEPATH__",
+	}
 	if apiRoot != "" {
-		k8sAPIRoot = fmt.Sprintf("\n          \"k8s_api_root\":\"%s\",", apiRoot)
+		kubernetes["k8s_api_root"] = apiRoot
 	}
+	calicoPlugin["kubernetes"] = kubernetes
 
-	var externalDataplane string = ""
 	if c.vppDataplaneEnabled() {
-		externalDataplane = `,
-      "dataplane_options": {
-        "type": "grpc",
-        "socket": "unix:///var/run/calico/cni-server.sock"
-      }`
+		calicoPlugin["dataplane_options"] = map[string]interface{}{
+			"type":   "grpc",
+			"socket": "unix:///var/run/calico/cni-server.sock",
+		}
+	}
+
+	return calicoPlugin
+}
+
+func (c *nodeComponent) createBandwidthPlugin() map[string]interface{} {
+	// bandwidth plugin
+	var bandwidthPlugin = map[string]interface{}{
+		"type":         "bandwidth",
+		"capabilities": map[string]bool{"bandwidth": true},
+	}
+
+	return bandwidthPlugin
+}
+
+func (c *nodeComponent) createPortmapPlugin() map[string]interface{} {
+	// Determine portmap configuration to use.
+	portmapPlugin := map[string]interface{}{
+		"type": "portmap",
+		"snat": true,
+		"capabilities": map[string]bool{
+			"portMappings": true,
+		},
 	}
 
-	// Build the CNI configuration json.
+	return portmapPlugin
+}
+
+// nodeCNIConfigMap returns a config map containing the CNI network config to be installed on each node.
+// Returns nil if no configmap is needed.
+func (c *nodeComponent) nodeCNIConfigMap() *corev1.ConfigMap {
+
```

Comment:
very much a nit, but this newline can be removed

Reply:
Done.